dimod/core/initialized.py

```diff
@@ -106,28 +106,31 @@ def parse_initial_states(self, bqm,
             initial_states_variables = list(bqm.variables)
             initial_states_vartype = bqm.vartype
         else:
-            initial_states_array, initial_states_variables = \
-                as_samples(initial_states)
-
             # confirm that the vartype matches and/or make it match
             if isinstance(initial_states, SampleSet):
                 initial_states_vartype = initial_states.vartype
             else:
                 # check based on values, defaulting to match the current bqm
-                initial_states_vartype = infer_vartype(initial_states_array) or bqm.vartype
+                initial_states_vartype = infer_vartype(initial_states) or bqm.vartype
+
+            # only create a copy of the initial states if vartype mismatch
+            copy = initial_states_vartype != bqm.vartype
```

Comment:
So, the sampler will sometimes receive a copy, sometimes the original sampleset. It will never know if it's safe to modify it in place.

I would rather:
- always copy (bad for perf), or
- **provide an explicit control argument**, something along the lines of `copy_only_if_necessary=True` (`False` meaning copy always).

Reply:
Yes, that's true. I'm leaning towards the second option you listed, but maybe `copy_always` instead, or just `copy` (matching the `copy` arg in `as_samples`, which always copies when `copy=True` and copies only if necessary when `False`).

I'm not sure what should be the default behaviour though -> seems like it would be less confusing if the default were `copy_always/copy=True`.

Reply:
I would default to the more performant option (copy only if needed), but would enable the developer to choose (sampler developer, so not a huge crowd).

Reply:
Added a `copy_always` parameter that defaults to `False`.